### PR TITLE
[FLINK-19796][table] Explicit casting should be made if the type of a…

### PR DESCRIPTION
**Pull request by cshuo**
…n element in `ARRAY` not equals with the derived component type

## What is the purpose of the change

This pull request checks whether types of elements in `ARRAY` equal with the derived component type during validating `ARRAY` sqlNode, and makes explicit casting if not equal.

## Brief change log
  - Add a new SqlFunction for `ARRAY` operator.

## Verifying this change
  - add plan test in CalcTest.
  - add IT test in CalcITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by cshuo** [comment]
CC @JingsongLi plz have a look

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            return null;
+        }
+
+        // explicit cast elements to component type if they are not same
+        if (opBinding instanceof SqlCallBinding) {
+            SqlCall call = ((SqlCallBinding) opBinding).getCall();
+            List<RelDataType> operandTypes = opBinding.collectOperandTypes();
+            List<SqlNode> operands = call.getOperandList();
+            int idx = -1;
+            for (RelDataType opType : operandTypes) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java:53]: Use for i ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                call.setOperand(idx, castTo(operands.get(idx), type));
+            }
+        }
+
+        return SqlTypeUtil.createArrayType(opBinding.getTypeFactory(), type, false);
+    }
+
+    private SqlNode castTo(SqlNode node, RelDataType type) {
+        return SqlStdOperatorTable.CAST.createCall(
+                SqlParserPos.ZERO,
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java:67]: I this OK?

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks @cshuo for your contribution, can you check map and multiset too?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java (hunk 2 above)
Comment (cshuo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/SqlArrayFunction.java]: Yes, it's ok to use ZERO.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/utils/SqlValidatorUtils.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * When the element element does not equal with the component type, making explicit casting.
+     *
+     * @param evenType derived type for element with even index
+     * @param oddType derived type for element with odd index
+     * @param sqlCallBinding description of call
+     */
+    private static void adjustTypeForMultisetConstructor(
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/utils/SqlValidatorUtils.java:59]: multise?

**Review by JingsongLi — APPROVED** [review]
Looks good to me.

**Pull request merged**